File: tools/blender-check.py
"""
Blender экспортын шалгалт — 2-р алхам: үүсгэсэн .py файлуудыг ЖИНХЭНЭ Blender дотор
ажиллуулж, камер бай руугаа харж байгааг Blender-ийн өөрийн математикаар шалгана.

    node tools/blender-check.mjs                                  # 1-р алхам
    blender -b -P tools/blender-check.py -- .blender-check        # 2-р алхам

`pip install bpy` хийсэн бол Blender-гүйгээр ч ажиллана:
    python3 tools/blender-check.py .blender-check
"""
import json
import math
import os
import runpy
import sys

import bpy
from bpy_extras.object_utils import world_to_camera_view
from mathutils import Vector

EPS_ANGLE = 1e-4      # градус
EPS_MM = 1e-3
EPS_UV = 1e-5


def fcurves(id_data):
    """Blender 4.4+ slotted actions болон хуучин хувилбар хоёуланг дэмжинэ."""
    ad = getattr(id_data, "animation_data", None)
    if not ad or not ad.action:
        return []
    act = ad.action
    try:
        if hasattr(act, "layers") and len(act.layers):
            cb = act.layers[0].strips[0].channelbag(ad.action_slot)
            if cb:
                return list(cb.fcurves)
    except Exception:
        pass
    try:
        return list(act.fcurves)
    except Exception:
        return []


def roll_deg(matrix):
    """Камерын өөрийн +Z тэнхлэгээр хэмжсэн roll (хөтөч дэх утгатай ижил конвенц)."""
    rot = matrix.to_3x3()
    fwd = (rot @ Vector((0, 0, -1))).normalized()
    up = (rot @ Vector((0, 1, 0))).normalized()
    world_up = Vector((0, 0, 1))
    if abs(fwd.dot(world_up)) > 0.9999:          # яг дээрээс/доороос харж байвал roll утгагүй
        return 0.0
    ref = (world_up - world_up.dot(fwd) * fwd).normalized()
    return math.degrees(math.atan2(ref.cross(up).dot(-fwd), ref.dot(up)))


def run_case(folder, case):
    fails = []

    def need(ok, msg):
        if not ok:
            fails.append(msg)

    bpy.ops.wm.read_homefile(use_empty=True)
    runpy.run_path(os.path.join(folder, case["file"]), run_name="__main__")

    scene = bpy.context.scene
    cam = bpy.data.objects.get("ShotCam")
    target = bpy.data.objects.get("CAM_TARGET")
    cam_data = bpy.data.cameras.get("ShotCam")
    need(cam and target and cam_data, "ShotCam / CAM_TARGET үүсээгүй")
    if fails:
        return fails

    need(scene.camera is cam, "идэвхтэй камер ShotCam болоогүй")
    need(scene.render.fps == case["fps"], f"fps {scene.render.fps} ≠ {case['fps']}")
    need(scene.frame_start == case["frame_start"] and scene.frame_end == case["frame_end"],
         f"фреймийн муж {scene.frame_start}-{scene.frame_end} ≠ {case['frame_start']}-{case['frame_end']}")
    res = [scene.render.resolution_x, scene.render.resolution_y]
    need(res == case["res"], f"нягтрал {res} ≠ {case['res']}")
    need(cam_data.sensor_fit == "VERTICAL" and abs(cam_data.sensor_height - 24.0) < EPS_MM,
         "сенсор VERTICAL / 24mm биш — линз хөтөчтэй таарахгүй")

    depsgraph = bpy.context.evaluated_depsgraph_get()
    for k in case["keys"]:
        scene.frame_set(k["frame"])
        depsgraph.update()
        cam_eval = cam.evaluated_get(depsgraph)
        tgt_world = target.evaluated_get(depsgraph).matrix_world.translation

        want = Vector(k["target"])
        need((tgt_world - want).length < 1e-4,
             f"[{k['frame']}] бай {tuple(round(v, 3) for v in tgt_world)} ≠ {tuple(round(v, 3) for v in want)}")

        uv = world_to_camera_view(scene, cam_eval, tgt_world)
        need(abs(uv.x - .5) < EPS_UV and abs(uv.y - .5) < EPS_UV,
             f"[{k['frame']}] камер бай руугаа харахгүй байна — кадрын байрлал "
             f"({uv.x:.5f}, {uv.y:.5f}), төв нь (0.5, 0.5)")
        need(abs(uv.z - k["distance"]) < 1e-3,
             f"[{k['frame']}] бай хүртэлх зай {uv.z:.3f}м ≠ {k['distance']}м")

        lens = cam_data.evaluated_get(depsgraph).lens
        need(abs(lens - k["lens"]) < EPS_MM, f"[{k['frame']}] линз {lens:.3f}mm ≠ {k['lens']:.3f}mm")

        got_roll = roll_deg(cam_eval.matrix_world)
        need(abs(got_roll - k["roll_deg"]) < 1e-3,
             f"[{k['frame']}] roll {got_roll:.3f}° ≠ {k['roll_deg']:.3f}°")

    for holder, label in ((cam, "камер"), (cam_data, "линз")):
        modes = {kp.interpolation for fc in fcurves(holder) for kp in fc.keyframe_points}
        need(modes == {case["interpolation"]},
             f"{label}ын интерполяц {modes or 'тавигдаагүй'} ≠ {{'{case['interpolation']}'}}")

    layout = bpy.data.collections.get("1stStudio_Layout")
    names = sorted(o.name for o in layout.objects) if layout else []
    want_names = sorted(case["subjects"] + case["props"])
    need(names == want_names, f"тайзны объектууд {names} ≠ {want_names}")

    text = bpy.data.texts.get("1stStudio_Prompt")
    need(text is not None and text.as_string().rstrip("\n") == case["prompt"].rstrip("\n"),
         "промтын текст блок эх бичвэртэйгээ таарахгүй")

    return fails


def main():
    args = sys.argv[sys.argv.index("--") + 1:] if "--" in sys.argv else sys.argv[1:]
    folder = args[0] if args else ".blender-check"
    manifest = os.path.join(folder, "expected.json")
    if not os.path.exists(manifest):
        sys.exit(f"{manifest} алга. Эхлээд: node tools/blender-check.mjs --out {folder}")

    cases = json.load(open(manifest, encoding="utf-8"))
    print(f"Blender {bpy.app.version_string} | {len(cases)} тест | {folder}\n")

    bad = 0
    for case in cases:
        try:
            fails = run_case(folder, case)
        except Exception as exc:                      # ачаалахгүй .py ч бас унал
            fails = [f"{type(exc).__name__}: {exc}"]
        if fails:
            bad += 1
            print(f"  ✘ {case['id']}")
            for f in fails:
                print(f"      {f}")
        else:
            print(f"  ✔ {case['id']}")

    print(f"\n{len(cases) - bad}/{len(cases)} тест давлаа.")
    sys.exit(1 if bad else 0)


main()

File: tools/blender-check.mjs
/**
 * Blender экспортын шалгалт — 1-р алхам: тест .py файлуудыг үүсгэнэ.
 *
 *   node tools/blender-check.mjs            → .blender-check/ дотор тест файлууд
 *   node tools/blender-check.mjs --out DIR
 *
 * app.js доторх ЖИНХЭНЭ экспортын кодыг (blenderKey / exportPY) тасдан авч
 * хөтөчгүйгээр ажиллуулна — өөр хуулбар байхгүй тул код өөрчлөгдвөл тест шууд мэдэрнэ.
 * Дараа нь 2-р алхам:  blender -b -P tools/blender-check.py -- .blender-check
 */
import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const require = createRequire(import.meta.url);
const THREE = require(join(root, "three.min.js"));
globalThis.THREE = THREE;

const outIdx = process.argv.indexOf("--out");
const outDir = outIdx > -1 ? process.argv[outIdx + 1] : join(root, ".blender-check");
mkdirSync(outDir, { recursive: true });

/* ── app.js-ээс хэрэгтэй хэсгүүдийг тасдаж авах ── */
const lines = readFileSync(join(root, "app.js"), "utf8").split("\n");
function cut(startRe, endRe, label) {
  const s = lines.findIndex(l => startRe.test(l));
  const e = lines.findIndex((l, i) => i > s && endRe.test(l));
  if (s < 0 || e < 0) throw new Error(`app.js дотроос "${label}" олдсонгүй — тестийг шинэчлэх хэрэгтэй`);
  return lines.slice(s, e).join("\n");
}
const code = [
  cut(/^function posOf\(s\)/, /^function applyCam/, "posOf"),
  cut(/^function applyCam\(cam, s, shake\)/, /^const cloneS/, "applyCam"),
  lines.find(l => l.startsWith("function lensMM")),
  cut(/Blender Python экспорт/, /22\. Toast/, "Blender экспорт"),
].join("\n");

/* ── Хөтчийн орчныг орлуулах хамгийн бага stub ── */
const shell = `
let fps = 24, fStart = 1, fEnd = 120, interp = 'smooth';
let keys = []; const people = [], props = [];
let ASPECT = 16 / 9, PROMPT = '', OUT = null;
function shotAspect() { return ASPECT; }
function toast(msg) { throw new Error('exportPY зогслоо: ' + msg); }
function stamp() { return 'check'; }
const $ = () => ({ value: '16:9', textContent: PROMPT });
function dl(name, data) { OUT = data; }
${code}
globalThis.__export = cfg => {
  ASPECT = cfg.aspect; fps = cfg.fps; fStart = cfg.fStart; fEnd = cfg.fEnd;
  interp = cfg.interp; keys = cfg.keys; PROMPT = cfg.prompt;
  people.length = 0; props.length = 0;
  cfg.people.forEach(p => people.push(p)); cfg.props.forEach(p => props.push(p));
  OUT = null; exportPY(); return OUT;
};`;
new Function(shell)();

/* ── Тест кейсүүд ── */
const V = (x, y, z) => new THREE.Vector3(x, y, z);
const key = (frame, theta, phi, radius, fov, roll, target) => ({ theta, phi, radius, fov, roll, target, frame });
const person = (x, z, ry, sc) => ({ position: V(x, 0, z), rotation: { y: ry }, scale: { x: sc } });
const prop = (kind, x, z, ry, sc) => ({ userData: { kind }, position: V(x, 0, z), rotation: { y: ry }, scale: { x: sc } });

const KEYS = [
  key(1,  0.0, Math.PI / 2.2, 6.0, 50,  0,     V(0, 1.2, 0)),
  key(48, 1.1, Math.PI / 2.6, 3.2, 35,  0.08,  V(0.5, 1.4, -1)),   // гурван тэнхлэгт эргэлт + roll
  key(96, 2.4, Math.PI / 3.0, 8.5, 85, -0.05,  V(-1, 1.0, 2)),
];
const PEOPLE = [person(0, 0, 0, 1), person(1.5, -2.2, 1.57, 1.05)];
const PROPS = [prop("car", -3, 1.2, 0.5, 1), prop("tree", 4, -4, 0, 2)];
const base = { aspect: 16 / 9, fps: 24, fStart: 1, fEnd: 96, interp: "smooth",
               keys: KEYS, people: PEOPLE, props: PROPS, prompt: "Шөнийн студи, dolly-in" };

const cases = [
  { id: "bezier",   cfg: base },
  { id: "linear",   cfg: { ...base, interp: "linear" } },
  { id: "const",    cfg: { ...base, interp: "const" } },
  { id: "vertical", cfg: { ...base, aspect: 9 / 16 } },
  { id: "empty",    cfg: { ...base, keys: [KEYS[0]], people: [], props: [] } },
  // промт нь хашилт, ар налуу зураас, мөр таслалт агуулсан хэцүү тохиолдол
  { id: "quotes",   cfg: { ...base, prompt: 'Камер: "dolly-in"\nЗам: C:\\render\\out\nТөгсгөл "' } },
];

const manifest = [];
for (const { id, cfg } of cases) {
  const py = globalThis.__export(cfg);
  if (!py) throw new Error(`[${id}] экспорт хоосон буцаалаа`);
  const file = `case-${id}.py`;
  writeFileSync(join(outDir, file), py, "utf8");
  const vertical = cfg.aspect < 1;
  manifest.push({
    id, file,
    fps: cfg.fps, frame_start: cfg.fStart, frame_end: cfg.fEnd,
    res: vertical ? [Math.round(1080 * cfg.aspect), 1080] : [1920, Math.round(1920 / cfg.aspect)],
    interpolation: { smooth: "BEZIER", linear: "LINEAR", const: "CONSTANT" }[cfg.interp],
    prompt: cfg.prompt,
    subjects: cfg.people.map((_, i) => `Subject_${String(i + 1).padStart(2, "0")}`),
    props: cfg.props.map((p, i) => `${p.userData.kind}_${String(i + 1).padStart(2, "0")}`),
    keys: cfg.keys.map(k => ({
      frame: k.frame,
      target: [k.target.x, -k.target.z, k.target.y],       // Blender (Z дээш) координат
      lens: 12 / Math.tan(k.fov * Math.PI / 360),
      roll_deg: (k.roll || 0) * 180 / Math.PI,
      distance: k.radius,
    })),
  });
  console.log(`  ✔ ${file.padEnd(18)} ${py.length} тэмдэгт`);
}
writeFileSync(join(outDir, "expected.json"), JSON.stringify(manifest, null, 2), "utf8");

console.log(`\n${cases.length} тест файл үүслээ: ${outDir}`);
console.log(`\nДараагийн алхам — Blender дээр шалгах:\n  blender -b -P tools/blender-check.py -- ${outDir}`);
